### PR TITLE
Document the process and get things ready

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

File: registry/DanTup_tiler.test [hunk 1]
```diff
@@ -0,0 +1,7 @@
+contact=ian@hixie.ch
```
Comment (tvolkert) [inline registry/DanTup_tiler.test]: Danny too?

File: registry/template.test [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+# not and will not be in GitHub, please contact ian@hixie.ch before
+# submitting a PR proposing to add such tests. We may make exceptions
+# if your tests are especially valuable.
+#
+# The checkout must always check out the exact same set of tests. Do
+# not check out the current master branch or some such; always fetch a
+# specific revision. You may regularly submit PRs that update the
+# revision being used.
+#
+# Each command is run in a separate shell.
```
Comment (tvolkert) [inline registry/template.test]: Specify which environment / shell.  e.g. Linux / bash.

File: registry/template.test [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#
+# The commands specified on the "fetch" lines must return the exit
+# code 0 or else the tests will be skipped entirely. The "git" program
```
Comment (tvolkert) [inline registry/template.test]: Will they be silently skipped?  If so, how can the developer know that their tests are running successfully short of checking the logs?

File: registry/template.test [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+# will be on the path.
```
Comment (tvolkert) [inline registry/template.test]: Anything else they can rely being on the path?  Any forbidden executables?  What about package managers (yum, rpm) or curl (which would introduce a dependency on something other than GitHub even if it's not where their code is hosted)?

File: registry/template.test [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+# must be at least two, and the first two must follow the pattern
+# below more or less exactly. Additional lines are allowed e.g. to
+# initialize and update submodules, but this should be rare. If this
+# is a problem, let us know in your PR.)
+fetch=git clone https://github.com/USERNAME/REPOSITORY.git tests
+fetch=git -C tests checkout 0123456789abcdef0123456789abcdef01234567
+
+# The "update" lines specify paths (relative to the newly created
+# "tests" directory) in which to run a flutter command to
+# automatically update the code to fix breaking changes. The command
```
Comment (tvolkert) [inline registry/template.test]: This is unclear.  It makes it sound like we have this magic command we'll run to fix their tests, and the "update" line specifies whence to run this command.

File: registry/template.test [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+#
+# If there are multiple directories, create a shell script in the
+# repository to enter each directory and run the tests in that
+# directory, returning non-zero as soon as any set of tests fails. You
+# can have multiple such scripts. In the example below, we run
+# "flutter test" in the root of the repository, a "more_tests.sh"
+# shell script from that same directory (the "./" indicates it isn't a
+# command to search for on the path), and the "extra_tests.dart"
+# program in the "dev" directory, run via "dart".
+test=flutter test
```
Comment (tvolkert) [inline registry/template.test]: Need we support anything other than `flutter test`, at least initially?

File: registry/template.test [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+test=./more_tests.sh
+test=dart dev/extra_tests.dart
+
+# To test your tests, check out the flutter/flutter repository and
+# then, from the root of that repository, run:
+#   bin/flutter; time bin/cache/dart-sdk/bin/dart dev/customer_testing/run_tests.dart --repeat=100 <path>
```
Comment (tvolkert) [inline registry/template.test]: Makes me wonder if we should create a `flutter heartbeat` or `flutter no-op` command that won't spit the usage to stdout...

File: README.md [hunk 8]
```diff
@@ -1,2 +1,95 @@
...
+`registry/template.test` to create a new file in the `registry/`
+directory. Fill in the fields and delete all the comments. Then,
+submit a PR with this new file.
+
+Tests must fulfill the following criteria to be added:
+
+* All the code must be available publicly on GitHub under a license
+  compatible with this effort.
+
+* Tests must be hermetic. For example, a test should not involve
```
Comment (tvolkert) [inline README.md]: What about the setup required to run a test (aside from cloning the repo)?

File: registry/template.test (hunk 2 above)
Comment (Hixie) [reply registry/template.test]: I guess technically it's not even a shell, I just spawn the process after splitting on spaces.

File: registry/DanTup_tiler.test (hunk 1 above)
Comment (Hixie) [reply registry/DanTup_tiler.test]: oops. replaced me with dantup.

File: registry/template.test (hunk 3 above)
Comment (Hixie) [reply registry/template.test]: We can control this, see https://github.com/flutter/flutter/pull/33140/files#diff-0f7067c31caedf7df2739c9b07efa122R161

I'm not sure what the default should be. If someone deletes their repo, I don't want us to start failing builds. I figure we'd make the cirrus tests for THIS repo not skip on failure, but have the main flutter repo skip on fetch failure.

File: registry/template.test (hunk 4 above)
Comment (Hixie) [reply registry/template.test]: \*shrug\*

We say the tests must be hermetic. I think it's up to the code reviewer to check that.

File: registry/template.test (hunk 5 above)
Comment (Hixie) [reply registry/template.test]: That is correct.

In practice it does nothing today.

File: registry/template.test (hunk 6 above)
Comment (Hixie) [reply registry/template.test]: \*shrug\*. I don't see why not.

File: README.md (hunk 8 above)
Comment (Hixie) [reply README.md]: like what?

Comment (tvolkert) [reply README.md]: Like downloading an Android SDK or a flutter_tools plugin (when such a thing exists)

**Comment by rrousselGit** [comment]
Hello!
I'm not sure if that matters but: I've read the readme and other files quite a few times, and I'm still not 100% sure that I understand what this repository is about.

Is it something that could help with https://github.com/flutter/flutter/issues/32474? 

It may be interesting to develop a bit more the description. 😄

**Comment by Hixie** [comment]
It would help us know when we are going to break someone, yes. And let us reach out.

File: README.md (hunk 8 above)
Comment (Hixie) [reply README.md]: The Android SDK should be provided by the test environment (still working on the details of that).
Plugins should be pinned by version, I guess. Maybe we should require pubspec.lock files?

Comment (tvolkert) [reply README.md]: Good point - we totally should.

File: .cirrus.yml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+task:
```
Comment (tvolkert) [inline .cirrus.yml]: /cc @fkorotkov for a review of whether this is canonical usage of `.cirrus.yml`

File: README.md [hunk 10]
```diff
@@ -1,2 +1,95 @@
-# tests
-Contributed tests for Flutter
+# Flutter Tests
+
+This repository contains references to tests that are run with every
+commit to Flutter to verify that no breaking changes have been
+introduced.
```
Comment (tvolkert) [inline README.md]: Specifically clarify that these tests are typically maintained by people outside of the Flutter team to give the Flutter team visibility into how their changes affect real-world developers using Flutter.

File: .cirrus.yml (hunk 9 above)
Comment (fkorotkov) [reply .cirrus.yml]: `matrix` looks invalid.

BTW name of this task will be `main`. If you want to change it you can either use `FOO_task` syntax or add `name: FOO` field to the `task`.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,95 @@
...
+
+* The tests must pass at the time they are contributed.
+
+
+## If a test is broken
+
+The point of these tests is to make sure we don't break existing code,
+while still being able to make improvements to Flutter APIs.
+
+If you find that a PR you have created causes one these tests to fail,
```
Comment (tvolkert) [inline README.md]: "... you have created in `flutter/flutter`"?

File: .cirrus.yml (hunk 9 above)
Comment (Hixie) [reply .cirrus.yml]: Is there documentation on this? I found https://cirrus-ci.org/guide/writing-tasks/ but it's really vague and doesn't actually list all the possible directives and where they're allowed and what the syntax is and so on.

Comment (fkorotkov) [reply .cirrus.yml]: There is `Task Naming` tip on this page. If your editor support JSON schemas from http://schemastore.org/json/ then you can get code assistance for `.cirrus.yml` files based on [`.cirrus.yml` schema](http://json.schemastore.org/cirrus).

Comment (Hixie) [reply .cirrus.yml]: That's a start, but it seems incomplete. For example, the schema doesn't seem to say what fields are required (e.g. it doesn't seem to say that "image" is required for "container", though if you omit "image" from "container" the bot complains). It also doesn't answer more fundamental questions like "what does it mean to specify both 'container' and 'windows_container'" or "what images can I use for a 'container'". :-(

Comment (fkorotkov) [reply .cirrus.yml]: Yeah, it's not possible to describe with the JSON schema standard. Create a feature request: https://github.com/cirruslabs/cirrus-ci-docs/issues/334

**Comment by Hixie** [comment]
(The bots are failing successfully in their current configuration.)

**Comment by 666Json** [comment]
- - - [ ] @PiNen>

**Pull request merged**